ProxySQL 2.0.10 crashing due to signal 11

ProxySQL 2.0.10 crashed with Signal 11.

ProxySQL version - 2.0.10-27-g5b31997
OS version - Ubuntu 16.04.6 LTS

**Proxysql error log :**

2021-11-09 13:00:01 MySQL_Thread.cpp:4583:process_all_sessions(): [WARNING] Closing unhealthy client connection 172.33.96.38:27526
Error: signal 11:
proxysql(_Z13crash_handleri+0x2d)[0x4ae7fd]
/lib/x86_64-linux-gnu/libc.so.6(+0x354c0)[0x7f5df8df24c0]
proxysql(_ZN12MySQL_Thread20process_all_sessionsEv+0x599)[0x4f3299]
proxysql(_ZN12MySQL_Thread3runEv+0x16e8)[0x4fb258]
proxysql(_Z24mysql_worker_thread_funcPv+0x7c)[0x4a740c]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba)[0x7f5df9e3b6ba]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x7f5df8ec44dd]
2021-11-09 13:00:06 main.cpp:1538:ProxySQL_daemonize_phase3(): [ERROR] ProxySQL crashed. Restarting!
2021-11-09 13:00:06 [INFO] ProxySQL version 2.0.10-27-g5b31997
2021-11-09 13:00:06 [INFO] ProxySQL SHA1 checksum: a9d97b741a0eb29deca65410a4e82319d1b4f2ce
2021-11-09 13:00:06 [INFO] Angel process started ProxySQL process 6805
Standard ProxySQL MySQL Logger rev. 2.0.0714 -- MySQL_Logger.cpp -- Mon Feb 24 16:38:57 2020
Standard ProxySQL Cluster rev. 0.4.0906 -- ProxySQL_Cluster.cpp -- Mon Feb 24 16:38:57 2020
Standard ProxySQL Statistics rev. 1.4.1027 -- ProxySQL_Statistics.cpp -- Mon Feb 24 16:38:57 2020
Standard ProxySQL HTTP Server Handler rev. 1.4.1031 -- ProxySQL_HTTP_Server.cpp -- Mon Feb 24 16:38:57 2020
2021-11-09 13:00:06 [INFO] Switching query logging format from 1 to 2
Standard ProxySQL Admin rev. 2.0.6.0805 -- ProxySQL_Admin.cpp -- Mon Feb 24 16:38:57 2020
2021-11-09 13:00:06 [INFO] ProxySQL SHA1 checksum: a9d97b741a0eb29deca65410a4e82319d1b4f2ce
Standard MySQL Threads Handler rev. 0.2.0902 -- MySQL_Thread.cpp -- Mon Feb 24 16:38:57 2020
Standard MySQL Authentication rev. 0.2.0902 -- MySQL_Authentication.cpp -- Mon Feb 24 16:38:57 2020
2021-11-09 13:00:06 [INFO] Dumping mysql_servers_incoming


**Core dump :**


GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/proxysql...done.
[New LWP 1705]
[New LWP 1694]
[New LWP 1696]
[New LWP 1697]
[New LWP 1700]
[New LWP 1698]
[New LWP 1701]
[New LWP 1699]
[New LWP 1702]
[New LWP 1703]
[New LWP 1711]
[New LWP 1706]
[New LWP 1712]
[New LWP 1713]
[New LWP 1735]
[New LWP 1736]
[New LWP 1704]
[New LWP 1707]
[New LWP 1740]
[New LWP 1709]
[New LWP 1822]
[New LWP 1826]
[New LWP 1858]
[New LWP 1708]
[New LWP 6800]
[New LWP 1741]
[New LWP 1743]
[New LWP 1739]
[New LWP 3490]
[New LWP 1738]
[New LWP 1742]
[New LWP 1859]
[New LWP 1737]
[New LWP 1710]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `proxysql -c /etc/proxysql.cnf -D /var/lib/proxysql'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  MySQL_Thread::process_all_sessions (this=this@entry=0x7f5df0000000) at MySQL_Thread.cpp:4487
4487	MySQL_Thread.cpp: No such file or directory.
[Current thread is 1 (Thread 0x7f5df61eb700 (LWP 1705))]
(gdb) bt full
#0  MySQL_Thread::process_all_sessions (this=this@entry=0x7f5df0000000) at MySQL_Thread.cpp:4487
        a = 0
        n = 37
        total_active_transactions_ = 0
        idle_maintenance_thread = false
        rc = <optimized out>
        __func__ = "process_all_sessions"
#1  0x00000000004fb258 in MySQL_Thread::run (this=this@entry=0x7f5df0000000) at MySQL_Thread.cpp:4327
        num_idles = <optimized out>
        ttw = <optimized out>
        maintenance_interval = <optimized out>
        n = <optimized out>
        rc = <optimized out>
        idle_maintenance_thread = <optimized out>
        __PRETTY_FUNCTION__ = "void MySQL_Thread::run()"
        __func__ = "run"
#2  0x00000000004a740c in mysql_worker_thread_func (arg=0x7f5df78a2f10) at main.cpp:743
        thread_attr = {__size = '\000' <repeats 17 times>, "\020", '\000' <repeats 37 times>, __align = 0}
        tmp_stack_size = 8388608
        mysql_thread = 0x7f5df78a2f10
        worker = 0x7f5df0000000
#3  0x00007f5df9e3b6ba in start_thread (arg=0x7f5df61eb700) at pthread_create.c:333
        __res = <optimized out>
        pd = 0x7f5df61eb700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140041537894144, 308842705401010586, 0, 140725267269407, 140041537894848, 140041515671552, -363992290258724454, -364007572692280934}, mask_was_saved = 0}}, 
          priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#4  0x00007f5df8ec44dd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
No locals.
(gdb) 
#0  MySQL_Thread::process_all_sessions (this=this@entry=0x7f5df0000000) at MySQL_Thread.cpp:4487
        a = 0
        n = 37
        total_active_transactions_ = 0
        idle_maintenance_thread = false
        rc = <optimized out>
        __func__ = "process_all_sessions"
#1  0x00000000004fb258 in MySQL_Thread::run (this=this@entry=0x7f5df0000000) at MySQL_Thread.cpp:4327
        num_idles = <optimized out>
        ttw = <optimized out>
        maintenance_interval = <optimized out>
        n = <optimized out>
        rc = <optimized out>
        idle_maintenance_thread = <optimized out>
        __PRETTY_FUNCTION__ = "void MySQL_Thread::run()"
        __func__ = "run"
#2  0x00000000004a740c in mysql_worker_thread_func (arg=0x7f5df78a2f10) at main.cpp:743
        thread_attr = {__size = '\000' <repeats 17 times>, "\020", '\000' <repeats 37 times>, __align = 0}
        tmp_stack_size = 8388608
        mysql_thread = 0x7f5df78a2f10
        worker = 0x7f5df0000000
#3  0x00007f5df9e3b6ba in start_thread (arg=0x7f5df61eb700) at pthread_create.c:333
        __res = <optimized out>
        pd = 0x7f5df61eb700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140041537894144, 308842705401010586, 0, 140725267269407, 140041537894848, 140041515671552, -363992290258724454, -364007572692280934}, mask_was_saved = 0}}, 
          priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#4  0x00007f5df8ec44dd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
No locals.

Dear Renecanno,

Still proxysql is crashing even after upgraded to the latest version  2.3.2-10-g8cd66cf.

OS version - Ubuntu 16.04.6 LTS
Proxysql version : 2.3.2-10-g8cd66cf.

**Proxysql error log  :** 

```
Error: signal 11:
proxysql(_Z13crash_handleri+0x3d)[0x5f7cbd]
/lib/x86_64-linux-gnu/libc.so.6(+0x354c0)[0x7ff34098c4c0]
/lib/x86_64-linux-gnu/libc.so.6(+0x14e295)[0x7ff340aa5295]
proxysql(_ZN14MySQL_Protocol16generate_pkt_rowEbPPvPjhiPmPPc+0x143)[0x6828c3]
proxysql(_ZN13MySQL_Session16SQLite3_to_MySQLEP14SQLite3_resultPciP14MySQL_Protocolbb+0x1cc)[0x664ccc]
proxysql(_Z21admin_session_handlerP13MySQL_SessionPvP10_PtrSize_t+0x1c4a)[0x6d420a]
proxysql(_ZN13MySQL_Session80handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY___not_mysqlER10_PtrSize_t+0x2f)[0x661adf]
proxysql(_ZN13MySQL_Session20get_pkts_from_clientERbR10_PtrSize_t+0x8f9)[0x67c799]
proxysql(_ZN13MySQL_Session7handlerEv+0xc7)[0x67a787]
proxysql(_Z11child_mysqlPv+0x3a2)[0x696e22]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba)[0x7ff3419d56ba]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x7ff340a5e4dd]
 ---- proxysql(_Z13crash_handleri+0x3d) [0x5f7cbd] : crash_handler(int)
 ---- proxysql(_ZN14MySQL_Protocol16generate_pkt_rowEbPPvPjhiPmPPc+0x143) [0x6828c3] : MySQL_Protocol::generate_pkt_row(bool, void**, unsigned int*, unsigned char, int, unsigned long*, char**)
 ---- proxysql(_ZN13MySQL_Session16SQLite3_to_MySQLEP14SQLite3_resultPciP14MySQL_Protocolbb+0x1cc) [0x664ccc] : MySQL_Session::SQLite3_to_MySQL(SQLite3_result*, char*, int, MySQL_Protocol*, bool, bool)
 ---- proxysql(_Z21admin_session_handlerP13MySQL_SessionPvP10_PtrSize_t+0x1c4a) [0x6d420a] : admin_session_handler(MySQL_Session*, void*, _PtrSize_t*)
 ---- proxysql(_ZN13MySQL_Session20get_pkts_from_clientERbR10_PtrSize_t+0x8f9) [0x67c799] : MySQL_Session::get_pkts_from_client(bool&, _PtrSize_t&)
 ---- proxysql(_ZN13MySQL_Session7handlerEv+0xc7) [0x67a787] : MySQL_Session::handler()
 ---- proxysql(_Z11child_mysqlPv+0x3a2) [0x696e22] : child_mysql(void*)
2021-11-18 09:45:04 main.cpp:1245:ProxySQL_daemonize_phase3(): [ERROR] ProxySQL crashed. Restarting!
2021-11-18 09:45:04 [INFO] ProxySQL version 2.3.2-10-g8cd66cf
2021-11-18 09:45:04 [INFO] ProxySQL SHA1 checksum: 3e37fdfe543b79b7b54099f2d88e741e38ac1597
2021-11-18 09:45:04 [INFO] Angel process started ProxySQL process 13325
```

**Coredump :**

```
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/proxysql...done.
[New LWP 13324]
[New LWP 9353]
[New LWP 9355]
[New LWP 9356]
[New LWP 9357]
[New LWP 9358]
[New LWP 9360]
[New LWP 9362]
[New LWP 9361]
[New LWP 9368]
[New LWP 9359]
[New LWP 9365]
[New LWP 9372]
[New LWP 9363]
[New LWP 9367]
[New LWP 9369]
[New LWP 9371]
[New LWP 9370]
[New LWP 9373]
[New LWP 9374]
[New LWP 9380]
[New LWP 9383]
[New LWP 9364]
[New LWP 9377]
[New LWP 9384]
[New LWP 9385]
[New LWP 9389]
[New LWP 9366]
[New LWP 9395]
[New LWP 9382]
[New LWP 9386]
[New LWP 9381]
[New LWP 9387]
[New LWP 9397]
[New LWP 9376]
[New LWP 9396]
[New LWP 9379]
[New LWP 10052]
[New LWP 9388]
[New LWP 10048]
[New LWP 9378]
[New LWP 8145]
[New LWP 10051]
[New LWP 10049]
[New LWP 27620]
[New LWP 10050]
[New LWP 10053]
[New LWP 9375]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `proxysql --idle-threads -c /etc/proxysql.cnf -D /var/lib/proxysql'.
****Program terminated with signal SIGSEGV, Segmentation fault.****
#0  0x00007ff340aa5295 in __memcpy_avx_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-avx-unaligned.S:254
254	../sysdeps/x86_64/multiarch/memcpy-avx-unaligned.S: No such file or directory.
[Current thread is 1 (Thread 0x7ff3289ff700 (LWP 13324))]
(gdb) bt full
#0  0x00007ff340aa5295 in __memcpy_avx_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-avx-unaligned.S:254
No locals.
#1  0x00000000006828c3 in memcpy (__len=44461305, __src=<optimized out>, __dest=<optimized out>) at /usr/include/x86_64-linux-gnu/bits/string3.h:53
No locals.
#2  write_encoded_length_and_string (string=<optimized out>, prefix=<optimized out>, len=<optimized out>, val=44461305, p=<optimized out>) at MySQL_Protocol.cpp:132
No locals.
#3  MySQL_Protocol::generate_pkt_row (this=this@entry=0x7ff327c1d0a0, send=send@entry=true, ptr=ptr@entry=0x0, len=len@entry=0x0, sequence_id=<optimized out>, colnums=<optimized out>, fieldslen=0x7ff327dfdd00, 
    fieldstxt=0x7ff327c0cb80) at MySQL_Protocol.cpp:975
        col = 13
        rowlen = <optimized out>
        size = <optimized out>
        _ptr = <optimized out>
        l = 107
#4  0x0000000000664ccc in MySQL_Session::SQLite3_to_MySQL (this=this@entry=0x7ff327c0f000, result=0x7ff327c6e000, error=error@entry=0x0, affected_rows=<optimized out>, myprot=0x7ff327c1d0a0, 
    in_transaction=in_transaction@entry=false, deprecate_eof_active=false) at MySQL_Session.cpp:6606
        r = <optimized out>
        nTrx = <optimized out>
        setStatus = 2
        p = 0x7ff327c0cb80
        l = 0x7ff327dfdd00
        __PRETTY_FUNCTION__ = "void MySQL_Session::SQLite3_to_MySQL(SQLite3_result*, char*, int, MySQL_Protocol*, bool, bool)"
        sid = <optimized out>
#5  0x00000000006d420a in admin_session_handler (sess=<optimized out>, _pa=0x7ff33f25cf00, pkt=<optimized out>) at ProxySQL_Admin.cpp:4782
        SPA = 0x7ff33f25cf00
        pa = 0x7ff33f25cf00
        needs_vacuum = true
        error = 0x0
        cols = 16
        affected_rows = 0
        run_query = <optimized out>
        resultset = 0x7ff327c6e000
        strA = <optimized out>
        strB = <optimized out>
        strAl = <optimized out>
        strBl = <optimized out>
        query = 0x7ff327c150f0 "select * from stats_mysql_processlist"
        query_length = 38
        query_no_space = <optimized out>
        query_no_space_length = <optimized out>
        __func__ = "admin_session_handler"
#6  0x0000000000661adf in MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY___not_mysql (this=this@entry=0x7ff327c0f000, pkt=...) at MySQL_Session.cpp:3201
        __PRETTY_FUNCTION__ = "void MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY___not_mysql(PtrSize_t&)"
#7  0x000000000067c799 in MySQL_Session::get_pkts_from_client (this=this@entry=0x7ff327c0f000, wrong_pass=@0x7ff3289fcfa3: false, pkt=...) at MySQL_Session.cpp:3670
        handler_ret = 0
        c = <optimized out>
        __func__ = "get_pkts_from_client"
        __PRETTY_FUNCTION__ = "int MySQL_Session::get_pkts_from_client(bool&, PtrSize_t&)"
#8  0x000000000067a787 in MySQL_Session::handler (this=this@entry=0x7ff327c0f000) at MySQL_Session.cpp:4254
        handler_ret = 0
---Type <return> to continue, or q <return> to quit---return
        prepared_stmt_with_no_params = false
        wrong_pass = false
        pkt = {ptr = 0x7ff327c150c0, size = 42}
        __PRETTY_FUNCTION__ = "int MySQL_Session::handler()"
        __func__ = "handler"
#9  0x0000000000696e22 in child_mysql (arg=<optimized out>) at ProxySQL_Admin.cpp:4935
        rc = <optimized out>
        thread_attr = {__size = '\000' <repeats 17 times>, "\020", '\000' <repeats 37 times>, __align = 0}
        tmp_stack_size = 8388608
        myarg = <optimized out>
        client = <optimized out>
        fds = {{fd = 892, events = 1, revents = 1}}
        nfds = 1
        rc = <optimized out>
        mysql_thr = 0x7ff327c64000
        sess = 0x7ff327c0f000
        myds = 0x7ff327c1d000
#10 0x00007ff3419d56ba in start_thread (arg=0x7ff3289ff700) at pthread_create.c:333
        __res = <optimized out>
        pd = 0x7ff3289ff700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140682335352576, 1151186366440590190, 0, 140682676653903, 8388608, 140682390033968, -1144049712614683794, -1144100277351298194}, mask_was_saved = 0}}, 
          priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#11 0x00007ff340a5e4dd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:74
No locals.
#12 0x0000000000000000 in ?? ()
```

--- a/include/MySQL_Protocol.h
+++ b/include/MySQL_Protocol.h
@@ -37,6 +37,16 @@ class MySQL_ResultSet {
 	MySQL_ResultSet();
 	void init(MySQL_Protocol *_myprot, MYSQL_RES *_res, MYSQL *_my, MYSQL_STMT *_stmt=NULL);
 	void init_with_stmt(MySQL_Connection *myconn);
+	/**
+	 * @brief Simple initialization of resulset of 'MySQL_ResultSet' without a resulset.
+	 * @details This initialization allows to reuse the logic from function 'generate_pkt_row3' for filling
+	 *   the resulset for later extracting the generated 'PtrSizeArray' via 'buffer_to_PSarrayOut' and
+	 *   'get_resultset'.
+	 *
+	 *   IMPORTANT-NOTE: Other member functions are not safe to be used after this initialization.
+	 * @param myproto Used to initialize internal 'MySQL_Protocol' field.
+	 */
+	void buffer_init(MySQL_Protocol* myproto);
 	~MySQL_ResultSet();
 	unsigned int add_row(MYSQL_ROWS *rows);
 	unsigned int add_row(MYSQL_ROW row);

--- a/lib/MySQL_Protocol.cpp
+++ b/lib/MySQL_Protocol.cpp
@@ -2488,6 +2488,16 @@ MySQL_ResultSet::MySQL_ResultSet() {
 	buffer = NULL;
 	//reset_pid = true;
 }
+
+void MySQL_ResultSet::buffer_init(MySQL_Protocol* myproto) {
+	if (buffer==NULL) {
+		buffer=(unsigned char *)malloc(RESULTSET_BUFLEN);
+	}
+
+	buffer_used=0;
+	myprot = myproto;
+}
+
 void MySQL_ResultSet::init(MySQL_Protocol *_myprot, MYSQL_RES *_res, MYSQL *_my, MYSQL_STMT *_stmt) {
 	PROXY_TRACE2();
 	transfer_started=false;

--- a/lib/MySQL_Session.cpp
+++ b/lib/MySQL_Session.cpp
@@ -6680,13 +6680,21 @@ void MySQL_Session::SQLite3_to_MySQL(SQLite3_result *result, char *error, int af
 
 		char **p=(char **)malloc(sizeof(char*)*result->columns);
 		unsigned long *l=(unsigned long *)malloc(sizeof(unsigned long *)*result->columns);
+
+		MySQL_ResultSet MyRS {};
+		MyRS.buffer_init(myprot);
+
 		for (int r=0; r<result->rows_count; r++) {
 		for (int i=0; i<result->columns; i++) {
 			l[i]=result->rows[r]->sizes[i];
 			p[i]=result->rows[r]->fields[i];
 		}
-		myprot->generate_pkt_row(true,NULL,NULL,sid,result->columns,l,p); sid++;
+			sid = myprot->generate_pkt_row3(&MyRS, NULL, sid, result->columns, l, p, 0); sid++;
 		}
+
+		MyRS.buffer_to_PSarrayOut();
+		MyRS.get_resultset(myds->PSarrayOUT);
+
 		myds->DSS=STATE_ROW;
 
 		if (deprecate_eof_active) {

--- a/test/tap/tests/reg_test_3690-admin_large_pkts-t.cpp
+++ b/test/tap/tests/reg_test_3690-admin_large_pkts-t.cpp
@@ -1,0 +1,203 @@
+/**
+ * @file reg_test_3690-admin_large_pkts-t.cpp
+ * @brief This test is a regression test for issue #3690.
+ * @details The test performs the following operations:
+ *   1. Creates a table in ProxySQL internal SQLite using an admin connection.
+ *   2. Fills this table with random length strings, varying from small to higher than 0xFFFFFF size.
+ *   3. Queries the generated rows, and checks that the received rows length matches the previously inserted
+ *      rows length.
+ *
+ * @date 2021-12-01
+ */
+
+#include <tuple>
+#include <vector>
+#include <string>
+#include <iostream>
+#include <random>
+
+#include <mysql.h>
+
+#include "json.hpp"
+#include "tap.h"
+#include "command_line.h"
+#include "utils.h"
+
+using std::string;
+using std::vector;
+
+string create_testing_table_query(int col_num) {
+	string test_table_query { "CREATE TABLE reg_test_3690_table (id INT" };
+
+	for (int i = 0; i < col_num; i++) {
+		test_table_query += ", v" + std::to_string(i) + " TEXT";
+	}
+
+	test_table_query += ")";
+
+	return test_table_query;
+}
+
+uint32_t TESTING_TABLE_COLUMNS = 10;
+
+vector<uint32_t> generate_random_row_lens(int cols_num) {
+	std::random_device rd;
+	std::default_random_engine gen(rd());
+	std::uniform_int_distribution<> elem_type_dst(0, 4);
+	std::uniform_int_distribution<> small_value_dst(1, 30);
+	vector<uint32_t> row_lens {};
+
+	for (int i = 0; i < cols_num; i++) {
+		int elem_type = elem_type_dst(gen);
+
+		if (elem_type == 0) {
+			row_lens.push_back(0);
+		} else if (elem_type == 1) {
+			row_lens.push_back(small_value_dst(gen));
+		} else if (elem_type == 2) {
+			row_lens.push_back(small_value_dst(gen) + 0xFFFFFF*2);
+		} else {
+			row_lens.push_back(small_value_dst(gen) + 0xFFFFFF);
+		}
+	}
+
+	return row_lens;
+}
+
+string generate_insert_query(const vector<uint32_t>& cols_lens) {
+	string random_insert_query { "INSERT INTO reg_test_3690_table (" };
+
+	for (uint32_t i = 0; i < cols_lens.size(); i++) {
+		random_insert_query += "v" + std::to_string(i);
+
+		if (i != cols_lens.size() - 1) {
+			random_insert_query += ",";
+		}
+	}
+
+	random_insert_query += ") VALUES (";
+
+	for (vector<uint32_t>::const_iterator it = cols_lens.begin(); it != cols_lens.end(); it++) {
+		if (*it == 0) {
+			random_insert_query += "\"\"";
+		} else {
+			random_insert_query += "printf('%.' || " + std::to_string(*it) + " || 'c', '*')";
+		}
+
+		if (std::next(it) != cols_lens.end()) {
+			random_insert_query += ",";
+		}
+	}
+
+	random_insert_query += ")";
+
+	return random_insert_query;
+}
+
+using row_act_exp_lens = std::tuple<uint32_t, vector<uint32_t>, vector<uint32_t>>;
+
+/**
+ * @brief Returns a vector holding the length of all the elements found for each row in the resulset.
+ */
+vector<vector<uint32_t>> fetch_rows_lens(MYSQL_RES* res) {
+	vector<vector<uint32_t>> rows {};
+
+	if (res == NULL) {
+		return rows;
+	}
+
+	MYSQL_ROW row = nullptr;
+	int num_fields = mysql_num_fields(res);
+
+	while ((row = mysql_fetch_row(res))) {
+		vector<uint32_t> row_vals {};
+
+		unsigned long *lengths;
+		lengths = mysql_fetch_lengths(res);
+
+		for(int i = 1; i < num_fields; i++) {
+			if (row[i]) {
+				row_vals.push_back(lengths[i]);
+			} else {
+				row_vals.push_back(0);
+			}
+		}
+
+		rows.push_back(row_vals);
+	}
+
+	return rows;
+}
+
+void match_row_lens(MYSQL_RES* t1_rows, const vector<vector<uint32_t>>& exp_rows_lens) {
+	vector<vector<uint32_t>> act_rows_lens { fetch_rows_lens(t1_rows) };
+	vector<row_act_exp_lens> result {};
+
+	for (uint32_t row_id = 0; row_id < exp_rows_lens.size(); row_id++) {
+		nlohmann::json j_exp_row_lens(exp_rows_lens[row_id]);
+		nlohmann::json j_act_row_lens(act_rows_lens[row_id]);
+
+		ok(
+			exp_rows_lens[row_id] == act_rows_lens[row_id],
+			"Actual rows lengths should match expected ones:\n"
+			" - Expected: '%s'\n - Actual: '%s'",
+			j_exp_row_lens.dump().c_str(), j_act_row_lens.dump().c_str()
+		);
+	}
+}
+
+uint32_t COLUMN_NUM = 10;
+uint32_t ROW_NUM = 10;
+
+int main(int argc, char** argv) {
+	CommandLine cl;
+
+	if (cl.getEnv()) {
+		diag("Failed to get the required environmental variables.");
+		return EXIT_FAILURE;
+	}
+
+	MYSQL* proxysql_admin = mysql_init(NULL);
+
+	// Initialize connections
+	if (!proxysql_admin) {
+		fprintf(stderr, "File %s, line %d, Error: %s\n", __FILE__, __LINE__, mysql_error(proxysql_admin));
+		return EXIT_FAILURE;
+	}
+
+	// Connnect to local proxysql
+	if (!mysql_real_connect(proxysql_admin, cl.host, cl.admin_username, cl.admin_password, NULL, cl.admin_port, NULL, 0)) {
+		fprintf(stderr, "File %s, line %d, Error: %s\n", __FILE__, __LINE__, mysql_error(proxysql_admin));
+		return EXIT_FAILURE;
+	}
+
+	// There should be a test for each inserted row
+	plan(ROW_NUM);
+
+	// Drop the testing table if exists
+	MYSQL_QUERY(proxysql_admin, "DROP TABLE IF EXISTS reg_test_3690_table");
+
+	// Create the testing table
+	std::string testing_table_query { create_testing_table_query(COLUMN_NUM) };
+	MYSQL_QUERY(proxysql_admin, testing_table_query.c_str());
+
+	// Insert the randomly sized rows
+	vector<vector<uint32_t>> inserted_rows {};
+	for (int i = 0; i < ROW_NUM; i++) {
+		auto row = generate_random_row_lens(COLUMN_NUM);
+		string insert_query = generate_insert_query(row);
+		MYSQL_QUERY(proxysql_admin, insert_query.c_str());
+
+		inserted_rows.push_back(row);
+	}
+
+	// Get the rows
+	MYSQL_QUERY(proxysql_admin, "SELECT * FROM reg_test_3690_table");
+	MYSQL_RES* select_res = mysql_store_result(proxysql_admin);
+	match_row_lens(select_res, inserted_rows);
+
+	// Cleanup the used table
+	MYSQL_QUERY(proxysql_admin, "DROP TABLE IF EXISTS reg_test_3690_table");
+
+	return exit_status();
+}

